RFC/src/RFC-0170_NetworkCommunicationProtocol.md

```diff
@@ -0,0 +1,301 @@
+ - automatically selecting a random subset or
+ - selecting a subset of neighbouring nodes with similar node ids. 
+
+ Maintaining communication channels are important and the following process can be followed in an attempt to keep peer connections alive:
+For an existing peer connection.
+When more than 30 minutes have passed since the last communication with that peer, then a heartbeat message should be sent to the peer in an attempt to keep the connection with the peer alive.
+If that specific peer connection is not important then a new peer can be selected from the local routing table.
+A new connection can then be established and maintained between the current node and the newly selected node.
+If that specific connection is important, such as with the connections between committee members, then the current CN or CC must wait and attempt to create a new connection with that same peer.
+If more than 90 minutes have passed since the last successful communication with the peer node, a new discovery request can be sent on the Tari communication network in an attempt to locate that peer again.
+Losing of a peer might happen in cases where the CN or CC went temporarily offline and their dynamic communication address changed, requiring the discovery process to be performed again before a direct P2P communication channel can be established.
+
+#### Functionality Required of Consensus Nodes
+
+- It MUST select a cryptographic key pair used for identification on the Tari Communication network.
+- A CN MAY request the peer addresses of CNs with similar node ids from other CNs to extend their local routing table. 
+- If a CN is a VN, then a node id MUST be obtained by registering on the Base layer.
+- If a CN is a BN, then a node id MUST be derived from the nodes identification public key.
+- A new CN MUST submit a joining request to the Tari communication network so that the nodes peer address can be added to the routing table of neighbouring peers in the network.
+- If a CN receives a new joining request with a similar node id (within a network selected threshold), then the peer address specified in the joining request MUST be added to its local routing table.
+- When a CN receives an encrypted message, the node MUST attempt to open the message.
+- When a CN receives an encrypted message that the node is unable to open, and the destination node id is known then the CN MUST forward it to all connected peers that have node ids that are closer to the destination.
+- When a CN receives an encrypted message that the node is unable to open and the destination node is unknown then the CN MUST forward the message to all connected peers.
+- A CN MUST have the ability to verify the content of unencrypted messages to limit the propagation of spam messages.
+- If an unencrypted message is received by the CN with a unspecified destination node id, then the node MUST verify the content of the message and forward the message to all connected peers.
```

Comment:
I'm still concerned that this opens the door to large-scale spam attacks. Are there TTLs on these messages? When do nodes know to stop passing the message around?

Reply:
Unencrypted message can be checked, if you receive junk message from your peer, either he didnt check it, or he created it. He gets blacklisted. 

Encrypted messages  can be a bit more specific. Perhaps the created must sign it, and if he creates too many messages he gets blacklisted?

Reply:
A CN should probably have a message storage list for recently distributed messages and should only send a new message to peers if it hasn't been sent before.

Reply:
For gossip this is a requirement anyway I think.